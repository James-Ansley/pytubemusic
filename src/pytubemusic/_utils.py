from datetime import datetime, timedelta
from itertools import pairwise
from typing import Iterator, Mapping


def to_delta(timestamp: str) -> timedelta:
    """Converts a H?:M:S timestamp string to a timedelta"""
    template = "00:00:00"
    timestamp = template[:-len(timestamp)] + timestamp
    return (
            datetime.strptime(timestamp, "%H:%M:%S")
            - datetime.strptime(template, "%H:%M:%S")
    )


def to_timestamp(delta: timedelta) -> str:
    """Converts a timedelta to a H:M:S timestamp string"""
    return (
        f"{delta.seconds // 3600:02d}"
        f":{(delta.seconds // 60) % 60:02d}"
        f":{delta.seconds % 60:02d}"
    )


def make_track_data(
        track_data: list[Mapping],
        end: timedelta,
) -> Iterator[Mapping]:
    """
    Adds track number metadata to a list of track data and ensures tracks have
    start and end timestamps

    :param track_data: A list of string maps of incomplete track data
    :param end: The length of the Album in seconds
    :return: Yields string maps of track data
    """
    track_data = track_data + [{"start": to_timestamp(end)}]
    for i, (t1, t2) in enumerate(pairwise(track_data), start=1):
        yield t1 | {
            "start": t1["start"],
            "end": t1.get("end", t2["start"]),
            "metadata": {**t1.get("metadata", {}), "track": i},
        }
