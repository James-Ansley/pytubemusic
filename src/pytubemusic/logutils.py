import contextlib
import functools
import inspect
import logging
import sys
from collections.abc import Iterable, Iterator
from textwrap import indent
from typing import Type

import cowsay

logger = logging.getLogger("pytubemusic")

INDENT_WIDTH = 2
_INDENT = 0

EType = Type[Exception] | tuple[Type[Exception], ...]


class PanicOn:
    def __init__(self, catch: EType, msg: str, handler=str):
        self.catch = catch
        self.msg = msg
        self.handler = handler

    def __enter__(self):
        ...

    def __exit__(self, exc_type, exc_val, exc_tb):
        if (
                isinstance(exc_val, self.catch)
                or (isinstance(self.catch, Iterable)
                    and any(isinstance(exc_val, e) for e in self.catch))
        ):
            message = cowsay.cowsay(
                self.msg, cow=cowsay.get_random_cow(), wrap_text=False
            )
            err = self.handler(exc_val)
            message = f"\x1b[31m{message}\n\n{err}\x1b[0m"
            logger.log(logging.ERROR, message)
            sys.exit(1)


@contextlib.contextmanager
def log_block(
        level=logging.INFO,
        on_enter: str = None,
        on_exit: str = None,
        on_error: str = None):
    """
    Logs entering, exiting, and errors in a block of code.
    Calls to log_block are on a "global stack" where each nested call is
    indented by a global INDENT_WIDTH.
    """
    global _INDENT, INDENT_WIDTH
    try:
        log_message(on_enter, [], {}, level)
        _INDENT += INDENT_WIDTH
        yield
        _INDENT -= INDENT_WIDTH
    except Exception as e:
        _INDENT -= INDENT_WIDTH
        log_message(on_error, [], {}, level)
        raise e
    else:
        log_message(on_exit, [], {}, level)


def log_call(
        level=logging.INFO,
        on_enter: str = None,
        on_exit: str = None,
        on_error: str = None,
):
    """
    Logs calls â€“ on a "global stack" where each nested call is indented
    by a global INDENT_WIDTH.
    """

    def decorator(func):
        arg_spec = inspect.getfullargspec(func)
        params = arg_spec.args

        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            global _INDENT, INDENT_WIDTH
            fmt_kwargs = dict(zip(params, args)) | kwargs
            fmt_args = [fmt_kwargs.get(n, "") for n in params]

            log_message(on_enter, fmt_args, fmt_kwargs, level)

            _INDENT += INDENT_WIDTH
            try:
                result = func(*args, **kwargs)
                _INDENT -= INDENT_WIDTH
            except Exception as e:
                _INDENT -= INDENT_WIDTH
                log_message(on_error, fmt_args, fmt_kwargs, level)
                raise e
            else:
                log_message(on_exit, fmt_args, fmt_kwargs, level)

            return result

        return wrapper

    return decorator


def log_message(msg, fmt_args, fmt_kwargs, level):
    if msg is not None:
        msg = indent(msg, " " * _INDENT).format(*fmt_args, **fmt_kwargs)
        logger.log(level, msg)


def log_iter(
        level=logging.INFO,
        on_each: str = None,
        on_enter: str = None,
        on_exit: str = None,
        start: int = 0,
):
    def iterator(data: Iterable) -> Iterator:
        global _INDENT
        if on_enter is not None and on_exit is not None:
            log_message(on_enter, [], {}, level)
            _INDENT += INDENT_WIDTH
        for i, e in enumerate(data, start=start):
            log_message(on_each, [e], {"i": i}, level)
            yield e
        if on_enter is not None and on_exit is not None:
            _INDENT -= INDENT_WIDTH
            log_message(on_exit, [], {}, level)

    return iterator
