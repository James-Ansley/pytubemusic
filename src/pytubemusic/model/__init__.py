import pydantic as _pydantic

from .base import *
from .album import *
from .cover import *
from .tags import *
from .track import *
from .track_data import *

type RootType = "Album | Track"
RootModel = _pydantic.RootModel[RootType]
