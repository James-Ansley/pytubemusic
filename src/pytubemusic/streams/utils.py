from contextlib import AbstractContextManager, contextmanager
from typing import IO


@contextmanager
def stream[T: IO](buff: T) -> AbstractContextManager[T]:
    try:
        yield buff
    finally:
        buff.seek(0)
