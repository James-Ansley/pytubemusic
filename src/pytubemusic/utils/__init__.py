from .logs import *
from .panics import *
