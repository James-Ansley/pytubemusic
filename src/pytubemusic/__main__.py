import logging
import tomllib
from pathlib import Path

from typer import Argument, Option, Typer

from .logutils import log_block, log_call
from .track import Track

app = Typer(add_completion=False)

logger = logging.getLogger("pytubemusic")
logger.setLevel(logging.INFO)
logger.addHandler(logging.StreamHandler())


@app.command(
    "album",
    help="Gets album tracks as individual files "
         "from the video associated with the album_data file",
)
@log_call(
    on_enter="Building album from '{album_data}'",
    on_exit="DONE",
    on_error="\x1b[31mProcessing Album Failed.\x1b[0m",
)
def make_album(
        album_data: Path = Argument(..., help="The album data"),
        out: Path = Option(
            Path("."),
            "--out", "-o",
            help="The directory album tracks will be written to. "
                 "Defaults to the album name",
        )
):
    with open(album_data, "rb") as f:
        album_data = tomllib.load(f)
    tracks = list(Track.from_album(**album_data))
    with log_block(on_enter="Exporting Tracks"):
        for track in tracks:
            track.export(out / track.album)


@app.command(
    "track",
    help="Gets a single audio file from the "
         "video associated with the track_data file",
)
@log_call(
    on_enter="Building track from '{track_data}'",
    on_exit="DONE",
    on_error="\x1b[31mProcessing Track Failed.\x1b[0m",
)
def make_track(
        track_data: Path = Argument(..., help="The track data"),
        out: Path = Option(
            Path("."),
            "--out", "-o",
            help="The directory album tracks will be written to. "
                 "Defaults to the cwd",
        )
):
    with open(track_data, "rb") as f:
        track_data = tomllib.load(f)
    track = Track.from_video(**track_data)
    track.export(out)


@app.command(
    "playlist",
    help="Gets tracks from the videos associated with the playlist",
)
@log_call(
    on_enter="Building album from '{playlist_data}'",
    on_exit="DONE",
    on_error="\x1b[31mProcessing Playlist Failed.\x1b[0m",
)
def make_playlist_album(
        playlist_data: Path = Argument(..., help="The track data"),
        out: Path = Option(
            Path("."),
            "--out", "-o",
            help="The directory album tracks will be written to. "
                 "Defaults to the cwd",
        )
):
    with open(playlist_data, "rb") as f:
        playlist_data = tomllib.load(f)
    tracks = list(Track.from_playlist(**playlist_data))
    with log_block(on_enter="Exporting tracks"):
        for track in tracks:
            track.export(out / track.album)


app()
