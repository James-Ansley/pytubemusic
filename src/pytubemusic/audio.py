import io
from collections.abc import Mapping
from datetime import timedelta
from typing import BinaryIO, Self

import backoff
from pydub import AudioSegment
from pytube import YouTube

from pytubemusic.utils import to_microseconds

__all__ = ["Audio"]


class Audio:
    def __init__(self, audio_data: AudioSegment, bitrate: int):
        self._audio_data = audio_data
        self.bitrate = bitrate

    @property
    def duration(self) -> timedelta:
        """A timedelta of the Audio length"""
        return timedelta(seconds=self._audio_data.duration_seconds)

    def snip(self, start: timedelta = None, end: timedelta = None) -> Self:
        """
        Returns a new audio object clamped between the two timestamps start
        and end. Both default to the start and end of the audio respectively.
        """
        if start is None and end is None:
            return self
        start = 0 if start is None else to_microseconds(start)
        end = to_microseconds(self.duration if end is None else end)
        return type(self)(self._audio_data[start:end], self.bitrate)

    def export(self, f: BinaryIO, metadata: Mapping[str, str], cover=None):
        """
        Writes the audio to the given BinaryIO in MP3 format with the given
        metadata.
        """
        self._audio_data.export(
            f,
            format="mp3",
            tags=metadata,
            parameters=["-b:a", f"{self.bitrate}"],
            cover=cover,
        )

    @classmethod
    @backoff.on_exception(backoff.expo, KeyError, max_tries=5)
    def from_url(cls, url: str) -> Self:
        """
        Downloads audio from the video associated with the URL

        :param url: The video URL
        :return: An Audio object
        """
        buffer = io.BytesIO()
        raw_audio = YouTube(url).streams.get_audio_only()
        raw_audio.stream_to_buffer(buffer)
        buffer.seek(0)
        audio_data = AudioSegment.from_file(buffer)
        return cls(audio_data, raw_audio.bitrate)
