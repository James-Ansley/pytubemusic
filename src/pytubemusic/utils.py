import contextlib
import urllib.request
from collections.abc import Iterable, Mapping
from datetime import datetime, timedelta
from itertools import chain, pairwise
from pathlib import Path, PurePath
from tempfile import NamedTemporaryFile
from typing import Any

from pytubemusic.logutils import PanicOn

__all__ = [
    "to_delta", "to_timestamp", "to_microseconds", "pathify",
    "set_ends", "merge_metadata", "get_cover", "pad", "File", "TrackData",
    "StrMap", "open_or_panic",
]

StrMap = Mapping[str, Any]
TrackData = Iterable[StrMap]

File = NamedTemporaryFile


def to_delta(timestamp: str | timedelta | None) -> timedelta | None:
    """Converts a H?:M:S.f? timestamp string to a timedelta"""
    if timestamp is None:
        return None
    elif isinstance(timestamp, timedelta):
        return timestamp

    if "." in timestamp:
        timestamp, microseconds = timestamp.split(".")
    else:
        microseconds = "00"
    template = "00:00:00"
    timestamp = template[:-len(timestamp)] + timestamp + "." + microseconds
    return (
            datetime.strptime(timestamp, "%H:%M:%S.%f")
            - datetime.strptime(template, "%H:%M:%S")
    )


def to_timestamp(delta: timedelta) -> str:
    """Converts a timedelta to a H:M:S.f timestamp string"""
    return (
        f"{int(delta.total_seconds() // 3600):02.0f}"
        f":{int(delta.total_seconds() // 60) % 60:02.0f}"
        f":{int(delta.total_seconds() % 60):02.0f}"
        f".{(delta.total_seconds() % 1) * 100:02.0f}"
    )


def to_microseconds(delta: timedelta) -> int:
    """Coverts a time delta to total_microseconds"""
    return int(delta.total_seconds() * 1000)


def pathify(root: PurePath, title: str, ext=".mp3") -> PurePath:
    """
    Replaces invalid characters in filenames and joins the params to a path
    with the given extension
    """
    return root / (title.replace("/", "\u2044") + ext)


def set_ends(track_data: TrackData) -> TrackData:
    """
    Fills in missing "end" timestamps setting them to the start of the next
    track. Returns a new mapping of track data.
    """
    for t1, t2 in pairwise(chain(track_data, [{"start": None}])):
        yield t1 | {
            "start": t1.get("start"),
            "end": t1.get("end", t2["start"]),
        }


def merge_metadata(track_data: TrackData, metadata: StrMap) -> TrackData:
    """
    Merges any album metadata with track-specific metadata.
    Track metadata overrides album metadata.
    """
    for data in track_data:
        yield {**data, "metadata": {**metadata, **data["metadata"]}}


def get_cover(cover, relative_to: Path) -> "None | NamedTemporaryFile":
    """Downloads an image into a named temporary file to be used as cover art"""
    match cover:
        case {"url": url}:
            f = NamedTemporaryFile(suffix='.jpg')
            img = urllib.request.urlopen(url).read()
            f.write(img)
            return f
        case {"file": path}:
            tmp = NamedTemporaryFile(suffix='.jpg')
            msg = f"Cannot find cover file"
            with open_or_panic(relative_to.parent / path, "rb", msg) as f:
                tmp.write(f.read())
            return tmp
        case None:
            return None
        case _:
            raise ValueError(f"Unrecognised Cover format: `{cover}`")


def pad(tracks, factory, length):
    tracks = list(tracks)
    tracks += [factory() for _ in range(length - len(tracks))]
    return tracks


@contextlib.contextmanager
def open_or_panic(file_name, mode, msg):
    with (PanicOn(OSError, msg), open(file_name, mode) as f):
        yield f
