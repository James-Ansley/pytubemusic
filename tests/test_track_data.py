from datetime import timedelta

from pytubemusic.model import *
from utils import test


@test
def a_single_can_be_flattened_into_track_data():
    single = Single(
        url="www.example.com",
        metadata=TrackTags(title="My Single Title"),
        cover=None,
        start="00:00:01",
        end="00:07:23",
    )
    expect = TrackData(
        metadata=Tags(title="My Single Title"),
        cover=None,
        track_parts=(
            AudioData(
                url="www.example.com",
                start=timedelta(seconds=1),
                end=timedelta(minutes=7, seconds=23),
            ),
        ),
    ),
    assert tuple(TrackData.from_track(single)) == expect


@test
def a_single_can_be_flattened_to_track_data_with_additional_metadata():
    single = Single(
        url="https://www.example.com",
        metadata=TrackTags(title="My Title"),
        cover=Url(href="https://www.example.com/photo.jpg"),
    )
    album_metadata = AlbumTags(album="My Album", composer="Mx Composer")
    expect = TrackData(
        metadata=Tags(
            title="My Title",
            album="My Album",
            composer="Mx Composer",
        ),
        cover="https://www.example.com/photo.jpg",
        track_parts=(AudioData(url="https://www.example.com"),),
    ),
    assert tuple(TrackData.from_track(single, album_metadata)) == expect


@test
def additional_metadata_does_not_override_flattened_single_metadata():
    single = Single(url="my_url", metadata=TrackTags(title="Title", track=1))
    album_metadata = AlbumTags(album="Album", title="Overridden", track=-1)
    expect = TrackData(
        metadata=Tags(title="Title", album="Album", track=1),
        cover=None,
        track_parts=(AudioData(url="my_url"),)
    ),
    assert tuple(TrackData.from_track(single, album_metadata)) == expect


@test
def flattened_singles_can_be_given_a_cover():
    single = Single(url="my_url", metadata=TrackTags(title="Title"))
    cover = Url(href="www.example.com")
    expect = TrackData(
        metadata=Tags(title="Title"),
        cover="https://www.example.com",
        track_parts=(AudioData(url="my_url"),),
    ),
    assert tuple(TrackData.from_track(single, cover=cover)) == expect


@test
def flattened_singles_will_keep_their_covers_even_if_another_is_given():
    single = Single(
        url="my_url",
        metadata=TrackTags(title="Title"),
        cover=File(path="/my_file.jpg"),
    )
    cover = Url(href="www.example.com")
    expect = TrackData(
        metadata=Tags(title="Title"),
        cover="file:///my_file.jpg",
        track_parts=(AudioData(url="my_url"),),
    ),
    assert tuple(TrackData.from_track(single, cover=cover)) == expect


@test
def split_tracks_can_be_flattened_to_multiple_track_data_objects():
    split = Split(
        url="www.example.com",
        tracks=(
            TrackStub(metadata=TrackTags(title="Track 1"), start="00:00:05"),
            TrackStub(metadata=TrackTags(title="Track 2"), start="00:00:10"),
            TrackStub(metadata=TrackTags(title="Track 3"), start="00:00:15"),
        )
    )
    expect = (
        TrackData(
            metadata=Tags(title="Track 1"),
            cover=None,
            track_parts=[
                AudioData(
                    url="www.example.com",
                    start=timedelta(seconds=5),
                    end=timedelta(seconds=10),
                )
            ],
        ),
        TrackData(
            metadata=Tags(title="Track 2"),
            cover=None,
            track_parts=[
                AudioData(
                    url="www.example.com",
                    start=timedelta(seconds=10),
                    end=timedelta(seconds=15),
                ),
            ],
        ),
        TrackData(
            metadata=Tags(title="Track 3"),
            cover=None,
            track_parts=[
                AudioData(
                    url="www.example.com",
                    start=timedelta(seconds=15),
                    end=None
                )
            ],
        ),
    )
    assert tuple(TrackData.from_track(split)) == expect


@test
def a_split_track_containing_a_single_track_can_be_flattened():
    split = Split(
        url="www.example.com",
        tracks=(
            TrackStub(metadata=TrackTags(title="Track 1"), start="00:00:05"),
        )
    )
    expect = (
        TrackData(
            metadata=Tags(title="Track 1"),
            cover=None,
            track_parts=[
                AudioData(url="www.example.com", start=timedelta(seconds=5)),
            ],
        ),
    )
    assert tuple(TrackData.from_track(split)) == expect


@test
def a_playlist_can_be_flattened_into_track_data():
    playlist = Playlist(
        url="www.example.com",
        cover=Url(href="www.example.com/pic.jpeg"),
        tracks=(
            TrackStub(
                metadata=TrackTags(title="My Title 1"),
                start=timedelta(seconds=1),
                end=timedelta(minutes=1, seconds=23),
            ),
            "DROP",
            TrackStub(
                metadata=TrackTags(title="My Title 3"),
                start=timedelta(minutes=5, seconds=56),
                end=timedelta(minutes=10, seconds=10),
            ),
        )
    )
    expect = (
        TrackData(
            metadata=Tags(title="My Title 1"),
            cover="https://www.example.com/pic.jpeg",
            track_parts=(PlaylistAudioData(
                url="www.example.com",
                index=0,
                start=timedelta(seconds=1),
                end=timedelta(minutes=1, seconds=23),
            ),)
        ),
        TrackData(
            metadata=Tags(title="My Title 3"),
            cover="https://www.example.com/pic.jpeg",
            track_parts=(PlaylistAudioData(
                url="www.example.com",
                index=2,
                start=timedelta(minutes=5, seconds=56),
                end=timedelta(minutes=10, seconds=10),
            ),)
        ),
    )
    assert tuple(TrackData.from_track(playlist)) == expect


@test
def a_merge_playlist_can_be_flattened_into_track_data():
    playlist = MergePlaylist(
        url="www.example.com",
        cover=Url(href="www.example.com/pic.jpeg"),
        metadata=TrackTags(title="My Title 1"),
        tracks=(
            TimeStub(
                start=timedelta(seconds=1),
                end=timedelta(minutes=1, seconds=23),
            ),
            "DROP",
            TimeStub(
                start=timedelta(minutes=5, seconds=56),
                end=timedelta(minutes=10, seconds=10),
            ),
        )
    )
    expect = (
        TrackData(
            metadata=Tags(title="My Title 1"),
            cover="https://www.example.com/pic.jpeg",
            track_parts=(
                PlaylistAudioData(
                    url="www.example.com",
                    index=0,
                    start=timedelta(seconds=1),
                    end=timedelta(minutes=1, seconds=23),
                ),
                PlaylistAudioData(
                    url="www.example.com",
                    index=2,
                    start=timedelta(minutes=5, seconds=56),
                    end=timedelta(minutes=10, seconds=10),
                ),
            )
        ),
    )
    assert tuple(TrackData.from_track(playlist)) == expect


@test
def a_merge_can_be_flattened_into_track_data():
    playlist = Merge(
        metadata=TrackTags(title="My Title 1"),
        cover=Url(href="www.example.com/pic.jpeg"),
        tracks=(
            AudioStub(
                url="www.example.com/1",
                start=timedelta(seconds=1),
                end=timedelta(minutes=1, seconds=23),
            ),
            AudioStub(
                url="www.example.com/2",
                start=timedelta(minutes=5, seconds=56),
                end=timedelta(minutes=10, seconds=10),
            ),
        )
    )
    expect = (
        TrackData(
            metadata=Tags(title="My Title 1"),
            cover="https://www.example.com/pic.jpeg",
            track_parts=(
                AudioData(
                    url="www.example.com/1",
                    start=timedelta(seconds=1),
                    end=timedelta(minutes=1, seconds=23),
                ),
                AudioData(
                    url="www.example.com/2",
                    start=timedelta(minutes=5, seconds=56),
                    end=timedelta(minutes=10, seconds=10),
                ),
            )
        ),
    )
    assert tuple(TrackData.from_track(playlist)) == expect