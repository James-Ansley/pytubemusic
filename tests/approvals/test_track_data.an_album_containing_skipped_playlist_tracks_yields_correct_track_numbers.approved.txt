(TrackData(metadata=Tags(title='Track 1'
                         album='My Album'
                         track=1),
           cover='https://www.example.com/pic.jpeg',
           track_parts=(PlaylistAudioData(url='www.example.com',
                                          index=0,
                                          start=None,
                                          end=None),)),
 TrackData(metadata=Tags(title='Track 2'
                         album='My Album'
                         track=2),
           cover='https://www.example.com/pic.jpeg',
           track_parts=(PlaylistAudioData(url='www.example.com',
                                          index=2,
                                          start=None,
                                          end=None),)))
